framework/transport/ieee1905_transport/include/mapf/transport/ieee1905_transport_messages.h

```diff
@@ -127,49 +137,108 @@ class CmduXxMessage : public Message {
+    /**
+     * Maximal number of types for a single subscribe/unsubscribe message
+     */
+    static constexpr int MAX_SUBSCRIBE_TYPES = 64;
+
+    /**
+     * The type of the request
+     */
+    enum class ReqType : uint8_t {
+        INVALID     = 0,
+        SUBSCRIBE   = 1,
+        UNSUBSCRIBE = 2,
+    };
 
-    virtual const std::string topic_prefix() const override { return kTopicPrefix; }
+    /**
+     * Message type to subscribe/unsubscribe
+     */
+    union MsgType {
+        struct {
+            uint32_t internal : 1;        // 0 - external [cmdu], 1 - internal
+            uint32_t vendor_specific : 1; // 0 - easymesh, 1 - vendor specific
+            uint32_t reserved : 14;       // unused
+            uint32_t type : 16;           // message type (opcode)
+        } bits;
+        uint32_t value = 0;
+    };
 
-    static const std::string ieee1905_topic(uint16_t msg_type)
+    struct Metadata {
+        uint8_t version       = kVersion;
+        ReqType type          = ReqType::INVALID;
+        uint8_t msg_types_num = 0;
+        MsgType msg_types[MAX_SUBSCRIBE_TYPES];
+    };
+
+    explicit SubscribeMessage(std::initializer_list<Frame> frames = {})
+        : Message(uint32_t(Type::SubscribeMessage), frames)
     {
-        return build_topic(kTopicPrefix, ETH_P_1905_1, msg_type);
+        // maximum one frame is allowed (if none are given we will allocate one below)
+        mapf_assert(this->frames().size() <= 1);
+
+        if (this->frames().empty()) {
+            Message::Frame frame(sizeof(Metadata));
+            Add(frame);
+        } else if (this->frames().back().len() < sizeof(Metadata)) {
+            this->frames().back().set_size(sizeof(Metadata));
+        }
     }
 
-    static const std::string lldp_topic() { return build_topic(kTopicPrefix, ETH_P_LLDP, 0x0000); }
-};
+    Metadata *metadata() const
+    {
+        mapf_assert(!frames().empty());
 
-class CmduTxMessage : public CmduXxMessage {
-    using CmduXxMessage::CmduXxMessage; // inherit base class constructors
+        return (Metadata *)frames().back().data();
+    };
 
-public:
-    static const std::string kTopicPrefix;
+    virtual std::ostream &print(std::ostream &os) const
+    {
+        Message::print(os);
 
-    virtual const std::string topic_prefix() const override { return kTopicPrefix; }
+        std::stringstream ss;
+        Metadata *m = metadata();
+        os << " metadata:" << std::endl;
+        os << "  version       : " << static_cast<int>(m->version) << std::endl;
+        os << "  type          : " << static_cast<int>(m->type) << std::endl;
+        os << "  msg_types_num : " << std::dec << int(m->msg_types_num) << std::endl;
+        os << "  msg_types     : " << std::hex << std::setfill('0') << std::setw(4)
+           << m->msg_types[0].value;
+        for (int i = 1; i < m->msg_types_num; i++) {
+            os << ", " << m->msg_types[i].value;
+        }
+        os << std::endl;
+
+        return os << ss.str();
```

Comment:
No idea what's going on here, not my code :)
But you have a point... I'll remove `ss`.

Reply:
actually similar implementation use the `ss` and `return os << ss.str();` at the end (as you did)
so maybe instead of removing `ss` it is better to stream to the `ss` and only at end of function edit the `os`